### PR TITLE
[WIP] input space warping

### PR DESCRIPTION
**Pull request by stefanocereda**
As discussed in #506 this is not for merge, but just to discuss how to better implement this feature

### CODE REVIEW

**Comment by betatim** [comment]
Thanks for this @stefanocereda. Who do you think would be a good person to pull into this discussion (we should ping them)? Personally I don't know very much about the input space warping, so I can't give you any useful feedback :)

We discussed that having a `Transformer` would be a good way to fit in with what is currently being done in skopt and that that isn't very easy to do. I'm going to be the annoying person who pushes for us finding a way to implement this as a transformer so it can also be used by other surrogates 👮 

Possibly naive questions/statements to start the ball rolling:
* beta warping is similar to log transforming or normalizing a dimension?
* unlike normalizing a dimension beta warping has parameters that need adjusting during the fit
* there is no way for the surrogate model to reach the instance of `Space` that it is operating on, and so can't adjust the parameters of the beta warp.

---

One thing that might help steer the discussion is if you can make some inline comments in the diff with questions/comments/pointers. 800lines of diff look intimidating to anyone trying to review them, with some pointers it will look less scary :)

**Comment by stefanocereda** [comment]
Ok, this should be compared against the gpr regressor provided by scikit-learn, I will try to provide a shorter pull-request with just the relevant code.

Yes the idea is to have a "tunable transformer", therefore I do not think it would be feasible to implement it as a transformer if that would prevent us from tuning it.

I will provide a shorter commit so that it becomes easier to reason about it.

**Comment by stefanocereda** [comment]
I've refactored the commit. I've created a new file (gpr_betaWarp) which is a verbatim copy of gpr.py, except it includes a different GPR, you should not focus on this file.

The important changes are inside warped_gpr.py, which extends the gpr defined in sklearn. Since The modifications are in the middle of the methods, I had to repeat much code. The relevant changes are  highlighted with a # DIFFERENCE comment.

You can use this regressor with gp_minimize simply by adding a use_betaWarp=true parameter.

Let me know what you think!

**Comment by stefanocereda** [comment]
Maybe it would be feasible to implement the transformer as a kernel? In this way it could be easy to access the hyperparameters and tune them, but I wouldn't know how to apply them
